ProcessTree.cpp

```diff
@@ -0,0 +1,530 @@
+                break;
+        }
+    }
+}
+
+void ProcessTree::AddPnForkQueue(int pid, int ppid)
+{
+    struct ProcessQueueItem p = {ProcessQueueFork, ProcessTreeSource_pnotify, pid, ppid};
+    std::unique_lock<std::mutex> queue_push_lock(_queue_push_mutex);
+    _PnQueue.push(p);
+    queue_push_lock.unlock();
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    _queue_data_ready = true;
+    _queue_data.notify_one();
+}
+
+void ProcessTree::AddPnExecQueue(int pid)
+{
+    struct ProcessQueueItem p = {ProcessQueueExec, ProcessTreeSource_pnotify, pid};
+    std::unique_lock<std::mutex> queue_push_lock(_queue_push_mutex);
+    _PnQueue.push(p);
+    queue_push_lock.unlock();
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    _queue_data_ready = true;
+    _queue_data.notify_one();
+}
+
+void ProcessTree::AddPnExitQueue(int pid)
+{
+    struct ProcessQueueItem p = {ProcessQueueExit, ProcessTreeSource_pnotify, pid};
+    std::unique_lock<std::mutex> queue_push_lock(_queue_push_mutex);
+    _PnQueue.push(p);
+    queue_push_lock.unlock();
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    _queue_data_ready = true;
+    _queue_data.notify_one();
+}
+
+void ProcessTree::run()
+{
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    while (!IsStopping()) {
+        _queue_data.wait(queue_pop_lock, [&]{return _queue_data_ready;});
+        while (!_PnQueue.empty()) {
+            struct ProcessQueueItem p = _PnQueue.front();
+            _PnQueue.pop();
+            switch (p.type) {
+                case ProcessQueueFork:
+                    AddPid(p.pid, p.ppid);
+                    break;
+                case ProcessQueueExec:
+                    AddPid(p.pid);
+                    break;
+                case ProcessQueueExit:
+                    RemovePid(p.pid);
+                    break;
+                default:
+                    Logger::Error("Invalid ProcessQueueType");
+            }
+        }
+
+        // Check if it's time for routine pruning of stale pids
+        std::chrono::duration<double> elapsed_seconds = std::chrono::system_clock::now() - _last_clean_time;
+        if (elapsed_seconds.count() > CLEAN_PROCESS_INTERVAL) {
+            Clean();
+            _last_clean_time = std::chrono::system_clock::now();
+        }
+        _queue_data_ready = false;
+    }
+}
+
+/* Process event from pnotify (fork)
+   If the pid is not in our table then add it.
+   If the ppid is in our table, copy the data from that entry.
+*/
+void ProcessTree::AddPid(int pid, int ppid)
+{
+    std::unique_lock<std::mutex> process_write_lock(_process_write_mutex);
+    if (_processes.count(pid) == 0) {
+        std::shared_ptr<ProcessTreeItem> process = std::make_shared<ProcessTreeItem>(ProcessTreeSource_pnotify, pid, ppid);
+        if (ppid && _processes.count(ppid) > 0) {
+            process->_uid = _processes[ppid]->_uid;
```

Comment:
This is a lot or repetitive _processes[ppid] hash lookups, it would be more efficient to fetch _processes[ppid], once (e.g. "auto pproc = _processes[ppid]") then use pproc in all subsequent references.

Reply:
Yep. I'm sure I had a reason for this originally (and assumed the optimisation in the compiler would only do the look up once) but you're correct it is much simpler and likely more efficient to do the look up once; implemented.